HeishaMon/decode.cpp

```diff
@@ -0,0 +1,446 @@
+#include "decode.h"
+#include "commands.h"
+
+unsigned long nextalldatatime = 0;
+
+
+// Decode ////////////////////////////////////////////////////////////////////////////
+void decode_heatpump_data(char* data, DynamicJsonDocument &actData, PubSubClient &mqtt_client, void (*log_message)(char*)) {
+  char log_msg[256];
+  char mqtt_topic[256];
+  
+  if (millis() > nextalldatatime) {
+    actData.clear(); // clearing all actual data so everything will be updated and sent to mqtt
+    nextalldatatime = millis() + UPDATEALLTIME;
+  }
+
+  // TOP3 //  
+  int Power_State = (int)(data[4]);
+  String Power_State_string;
+  switch (Power_State & 0b11) { //probably only last two bits for Power dhw state
+    case 0b01:
+      Power_State_string = "0";
+      break;
+    case 0b10:
+      Power_State_string = "1";
+      break;
+    default:
+      Power_State_string = "-1";
+      break;
+  }
+  if ( actData["Power_State"] != Power_State_string ) {
+    actData["Power_State"] = Power_State_string;
+    sprintf(log_msg, "received Power state : %d (%s)", Power_State, Power_State_string.c_str()); log_message(log_msg);
+    sprintf(mqtt_topic, "%s/%s", mqtt_topic_base, "Power_State"); mqtt_client.publish(mqtt_topic, Power_State_string.c_str(), MQTT_RETAIN_VALUES);
+  }
+
+  // TOP4 //
+  int Mode_State = (int)(data[6]);
+  String Mode_State_string;
+  switch (Mode_State) {
+    case 82:
+      Mode_State_string = "0";
+      break;
+    case 83:
+      Mode_State_string = "1";
+      break;
+    case 89:
+      Mode_State_string = "2";
+      break;
+    case 97:
+      Mode_State_string = "3";
+      break;
+    case 98:
+      Mode_State_string = "4";
+      break;
+    case 99:
+      Mode_State_string = "5";
+      break;
+    case 105:
+      Mode_State_string = "6";
+      break;
+    default:
+      Mode_State_string = "-1";
+      break;
+  }
+  if ( actData["OpMode_State"] != Mode_State_string ) {
+    actData["OpMode_State"] = Mode_State_string;
+    sprintf(log_msg, "received OpMode state : %d (%s)", Mode_State, Mode_State_string.c_str()); log_message(log_msg);
+    sprintf(mqtt_topic, "%s/%s", mqtt_topic_base, "OpMode_State"); mqtt_client.publish(mqtt_topic, Mode_State_string.c_str(), MQTT_RETAIN_VALUES);
+  }
+
+  // TOP18 //
+  int quietpower_Mode_State = (int)(data[7]);
+  String Powerfull_Mode_State_string = "-1";
+  String Quiet_Mode_State_string = "-1";
+  switch (quietpower_Mode_State & 0b11111000) { // only interested in left most 5 bits for quiet state
```

Comment:
its more clear to read and maintain to have this two String defs (74,75) in the default case like the other TOPs